Jolt: Update to 5.4.0

mihe opened the pull request:
This bumps Jolt Physics from jrouwe/JoltPhysics@0373ec0dd762e4bc2f6acdb08371ee84fa23c6db (version 5.3.0) to jrouwe/JoltPhysics@036ea7b1d717b3e713ac9d8cbd47118fb9cd5d60 (version 5.4.0).

Several of the changes in this release have already been backported to our copy, so the release notes aren't going to be entirely applicable, but [here](https://github.com/jrouwe/JoltPhysics/releases/tag/v5.4.0) they are anyway.

This brings in a few bug fixes that haven't been backported, and in general makes our copy more "proper" I guess, by not relying on backports anymore.

(Leaving this as draft until I've tested it a bit.)

CC @jrouwe.

mihe committed:
Jolt: Update to 5.4.0

mihe commented:
I was a bit curious about jrouwe/JoltPhysics#1758. Would you say that's something people are likely to run into, or notice at all?

jrouwe commented:
> I was a bit curious about [jrouwe/JoltPhysics#1758](https://github.com/jrouwe/JoltPhysics/pull/1758). Would you say that's something people are likely to run into, or notice at all?

This only makes a difference on bad meshes and most likely doesn't have a huge impact. it won't do anything if enhanced edge detection is on.

mihe commented:
It looks like the version check in `RegisterTypesInternal` is not playing nice with Godot's CI build cache, as the Jolt source files in `thirdparty/jolt_physics` are being recompiled with the new `JPH_VERSION_ID` but `modules/jolt_physics/jolt_globals.cpp` is still using the old one, leading to `std::abort()` being called.

None of the feature bits are causing problems, so I'm assuming it's because `JPH_VERSION_MINOR` is mixed into `JPH_VERSION_ID`. The mismatch mask is `0b11100000000`, which I guess makes sense if the minor version went from 3 to 4.

There's probably some way of wiping the CI cache, but this will cause issues for anyone compiling with caching enabled downstream as well, so that surely can't be the fix.

I don't really understand how this can happen in the first place, short of SCons doing some overly aggressive caching. There's no way `modules/jolt_physics/jolt_globals.cpp` should not have been invalidated and recompiled, since it has to include `Jolt/Core/Core.h`, which has the new version number in it.

I would've thought this would invalidate them either way:

https://github.com/godotengine/godot/blob/9283328fe70d27f148c403943a283d93a9785beb/modules/jolt_physics/SCsub#L178-L179

mihe commented:
> the Jolt source files in `thirdparty/jolt_physics` are being recompiled with the new `JPH_VERSION_ID` but `modules/jolt_physics/jolt_globals.cpp` is still using the old one

It seems it's actually the other way around, because `Jolt/RegisterTypes.cpp` isn't being recompiled for some reason. That's frankly just as weird though.

jrouwe commented:
It looks like in `modules/jolt_physics/SCsub' the following line needs to be changed from:

```
env_jolt.Prepend(CPPEXTPATH=[thirdparty_dir])
```

to

```
env_jolt.Prepend(CPPPATH=[thirdparty_dir])
```

For me this makes all the dependencies show up when running scons with `--tree=prune`:

```
  | | | +-bin\obj\modules\module_jolt_physics.windows.editor.x86_64.lib
  | | | | +-bin\obj\thirdparty\jolt_physics\Jolt\RegisterTypes.windows.editor.x86_64.obj
  | | | | | +-thirdparty\jolt_physics\Jolt\RegisterTypes.cpp
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Factory.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\RTTI.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\TickCounter.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Jolt.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\CollisionDispatch.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\PhysicsMaterialSimple.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\BoxShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\CapsuleShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\ConvexHullShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\CylinderShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\EmptyShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\HeightFieldShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\MeshShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\MutableCompoundShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\OffsetCenterOfMassShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\PlaneShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\RotatedTranslatedShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\ScaledShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\SphereShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\StaticCompoundShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\TaperedCapsuleShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\TaperedCylinderShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\TriangleShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\SoftBody\SoftBodyShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\RegisterTypes.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\UnorderedMap.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Reference.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\StaticArray.h
  | | | | | +-thirdparty\jolt_physics\Jolt\ObjectStream\SerializableAttribute.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\ARMNeon.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Array.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Core.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\IssueReporting.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Memory.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Mat44.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Math.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Real.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Vec4.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\NarrowPhaseStats.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\Shape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\SubShapeID.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\ShapeCast.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\ShapeFilter.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\PhysicsMaterial.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\ConvexShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\PhysicsSettings.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Geometry\Plane.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Renderer\DebugRenderer.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\ByteBuffer.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Geometry\IndexedTriangle.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Geometry\Triangle.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\CompoundShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\DecoratedShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\Shape\ScaleHelpers.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\HalfFloat.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\SortReverseAndStore.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\HashTable.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Atomics.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\HashCombine.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\STLAllocator.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Mat44.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\MathTypes.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\DMat44.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\DVec3.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Vec4.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Float4.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Swizzle.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Color.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\NonCopyable.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\Result.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\StreamUtils.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\UnorderedSet.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Geometry\AABox.h
  | | | | | +-thirdparty\jolt_physics\Jolt\ObjectStream\SerializableObject.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Body\MassProperties.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\BackFaceMode.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\CollisionCollector.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\CollideShape.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Body\BodyID.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Float2.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\STLAlignedAllocator.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\FPException.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\BVec16.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Quat.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Vec3.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\DMat44.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\DVec3.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Double3.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Trigonometry.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\UVec4.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\StreamIn.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\StreamOut.h
  | | | | | +-thirdparty\jolt_physics\Jolt\ObjectStream\ObjectStream.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\ActiveEdgeMode.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Physics\Collision\CollectFacesMode.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Core\FPControlWord.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\BVec16.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Quat.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Vec3.inl
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\Float3.h
  | | | | | +-thirdparty\jolt_physics\Jolt\Math\UVec4.inl
  | | | | | +-C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\bin\HostX64\x64\cl.EXE
```

However, on my machine it still doesn't rebuild.

jrouwe commented:
I think I figured it out, I now added both lines:

```
env_jolt.Prepend(CPPEXTPATH=[thirdparty_dir])
env_jolt.Prepend(CPPPATH=[thirdparty_dir])
```

The first disables warnings in external headers, the 2nd makes scons search the correct path for dependencies.

The reason why it still isn't working is that I was passing `dev_build=yes` which has a special check here:

https://github.com/godotengine/godot/blob/b4472f4670a0b7fd34137d18f1e080e3319e3819/SConstruct#L536-L538

that enables the option `implicit_cache`. According to [this](https://scons.org/doc/4.8.1/HTML/scons-user/ch06s03.html) means that any changes to `CPPPATH` are ignored. 

I commented those 3 lines out locally and then it works. This should at least fix all the non `dev_build`'s. I'm not sure what the build cache actually caches. If it only caches output files then we may get lucky, if it also caches scons state then it may still ignore the `CPPPATH` changes.

mihe commented:
`env_jolt.Prepend(CPPPATH=[thirdparty_dir])` was apparently changed to be `env_jolt.Prepend(CPPEXTPATH=[thirdparty_dir])` in #104893.

Maybe @Repiteo or someone else in @godotengine/buildsystem can help us figure out why that seems to mess with SCons tracking of header dependencies in `thirdparty/jolt_physics`. Perhaps it's related to the nixpkgs issue mentioned in that PR?

mihe commented:
Just to unblock this I've changed the Jolt `SCsub` file to use `CPPPATH` again, instead of `CPPEXTPATH`, which seems to still be a thing with a few other third-party integrations (pending #108611). I've also added a single whitespace to `‎thirdparty/jolt_physics/Jolt/RegisterTypes.cpp` in order to force that file to be compiled.

However, this of course does nothing for whatever other unchanged Jolt source files that might possibly need recompiling due to some header dependency having changed. It only addresses the `std::abort` crash that's preventing the CI tests from passing. Either way, the dangers of `fast_unsafe` won't be resolved in this PR.

Also, even with this hack, this can still cause problems if the following happens:

1. Somebody updates Jolt to a newer version by removing all the files in `thirdparty/jolt_physics`, effectively reverting this single whitespace, assuming `Jolt/RegisterTypes.cpp` hasn't changed in some other way by that point.
2. Somebody tries to bisect some bug, while having `dev_build=yes` enabled (and thus `fast_unsafe=yes`), and ends up jumping between two commits where this whitespace change never happened.

In such a case you will most likely `std::abort` in Jolt's version check and scratch your head for a while wondering why.

mihe commented:
The projects that I usually smoke test with all seem to behave as expected, so I've taken this out of draft.

jrouwe commented:
> Somebody updates Jolt to a newer version by overwriting all the files in thirdparty/jolt_physics, effectively reverting this single whitespace in Jolt/RegisterTypes.cpp, assuming it hasn't changed in some other way by that point.

I may have found a solution to that [problem](https://github.com/jrouwe/JoltPhysics/commit/9e48d59beea6a10ee8dfed45216c2dec7e9a8918) :)

akien-mga commented:
The `CPPEXTPATH` change was reverted in #111331 so you should be able to drop the custom changes to Jolt to fix CI.

Calinou reviewed (approved):
Tested locally on several demo projects, it works as expected. Code looks good to me.

mihe commented:
Well, I don't understand how, but we ended up having a report of this version check crash anyway, which was then resolved with a clean build.

Since this will likely keep being a problem when people go to bisect with `dev_build=yes`, I've put up a PR to cherrypick the recent change to `Jolt/RegisterTypes.cpp`, just to have some kind of change done to it: #111408
